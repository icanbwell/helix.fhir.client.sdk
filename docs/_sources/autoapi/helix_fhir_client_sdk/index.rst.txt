:py:mod:`helix_fhir_client_sdk`
===============================

.. py:module:: helix_fhir_client_sdk


Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   exceptions/index.rst
   filters/index.rst
   graph/index.rst
   loggers/index.rst
   responses/index.rst
   validators/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   fhir_client/index.rst
   well_known_configuration/index.rst


