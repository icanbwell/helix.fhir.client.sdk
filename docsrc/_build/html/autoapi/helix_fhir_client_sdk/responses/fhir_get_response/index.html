

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>helix_fhir_client_sdk.responses.fhir_get_response &mdash; helix.fhir.client.sdk  documentation</title>
  

  
  <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/graphviz.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../../" src="../../../../_static/documentation_options.js"></script>
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/underscore.js"></script>
        <script src="../../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../../index.html" class="icon icon-home"> helix.fhir.client.sdk
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">API Reference</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">helix.fhir.client.sdk</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><code class="xref py py-mod docutils literal notranslate"><span class="pre">helix_fhir_client_sdk.responses.fhir_get_response</span></code></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../../_sources/autoapi/helix_fhir_client_sdk/responses/fhir_get_response/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-helix_fhir_client_sdk.responses.fhir_get_response">
<span id="helix-fhir-client-sdk-responses-fhir-get-response"></span><h1><a class="reference internal" href="#module-helix_fhir_client_sdk.responses.fhir_get_response" title="helix_fhir_client_sdk.responses.fhir_get_response"><code class="xref py py-mod docutils literal notranslate"><span class="pre">helix_fhir_client_sdk.responses.fhir_get_response</span></code></a><a class="headerlink" href="#module-helix_fhir_client_sdk.responses.fhir_get_response" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this headline">¶</a></h2>
<div class="section" id="classes">
<h3>Classes<a class="headerlink" href="#classes" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#helix_fhir_client_sdk.responses.fhir_get_response.FhirGetResponse" title="helix_fhir_client_sdk.responses.fhir_get_response.FhirGetResponse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FhirGetResponse</span></code></a></p></td>
<td><p>Class that encapsulates the response from FHIR server</p></td>
</tr>
</tbody>
</table>
<dl class="py class">
<dt class="sig sig-object py" id="helix_fhir_client_sdk.responses.fhir_get_response.FhirGetResponse">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">helix_fhir_client_sdk.responses.fhir_get_response.</span></span><span class="sig-name descname"><span class="pre">FhirGetResponse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">responses</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">error</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">access_token</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#helix_fhir_client_sdk.responses.fhir_get_response.FhirGetResponse" title="Permalink to this definition">¶</a></dt>
<dd><p>Class that encapsulates the response from FHIR server</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>url</strong> (<em>str</em>) – url that was being accessed</p></li>
<li><p><strong>responses</strong> (<em>str</em>) – response text</p></li>
<li><p><strong>error</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Any error returned by FHIR server</p></li>
<li><p><strong>access_token</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – access token that was used</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
</div>
</div>


           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Helix Team.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>