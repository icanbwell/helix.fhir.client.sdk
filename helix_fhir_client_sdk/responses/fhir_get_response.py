from typing import Optional


class FhirGetResponse:
    def __init__(
        self,
        url: str,
        responses: str,
        error: Optional[str],
        access_token: Optional[str],
        total_count: Optional[int],
        status: int,
    ) -> None:
        """
        Class that encapsulates the response from FHIR server

        :param url: url that was being accessed
        :param responses: response text
        :param error: Any error returned by FHIR server
        :param access_token: access token that was used
        :param total_count: count of total records that match the provided query.
                            Only set if include_total_count was set to avoid expensive operation by server.
        """
        self.url: str = url
        self.responses: str = responses
        self.error: Optional[str] = error
        self.access_token: Optional[str] = access_token
        self.total_count: Optional[int] = total_count
        self.status: int = status
