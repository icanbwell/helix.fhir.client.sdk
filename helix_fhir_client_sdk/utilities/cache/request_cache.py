import asyncio
from types import TracebackType
from typing import Optional, Type, Dict

from helix_fhir_client_sdk.fhir.fhir_bundle_entry import FhirBundleEntry
from helix_fhir_client_sdk.utilities.cache.request_cache_entry import RequestCacheEntry


class RequestCache:
    """
    This is a class that caches requests to the FHIR server using a weak value dictionary.
    It is used to avoid multiple requests to the FHIR server when doing a large number
    of requests for the same resource.
    """

    __slots__ = [
        "cache_hits",
        "cache_misses",
        "_cache",
        "_lock",
    ]

    def __init__(self) -> None:
        self.cache_hits: int = 0
        self.cache_misses: int = 0
        self._cache: Dict[str, RequestCacheEntry] = {}
        self._lock: asyncio.Lock = asyncio.Lock()

    async def __aenter__(self) -> "RequestCache":
        """
        This method is called when the RequestCache is entered into a context manager.
        It returns the RequestCache instance.
        """
        async with self._lock:
            self._cache.clear()
        return self

    async def __aexit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_value: Optional[BaseException],
        traceback: Optional[TracebackType],
    ) -> Optional[bool]:
        """
        This method is called when the RequestCache is exited from a context manager.
        It clears the cache.
        """
        async with self._lock:
            self._cache.clear()

        if exc_type is not None:
            print(f"An exception of type {exc_type} occurred with message {exc_value}")
            return False  # Propagate any exception that occurred

        return True

    async def get_async(
        self, *, resource_type: str, resource_id: str
    ) -> Optional[RequestCacheEntry]:
        """
        This method returns the cached data for a given URL, or None if the URL is not in the cache.

        :param resource_type: The resource type to get the cached data for.
        :param resource_id: The resource id to get the cached data for.
        :return: The cached data for the given resource type and resource id, or None if the data is not in the cache.
        """
        key: str = f"{resource_type}/{resource_id}"

        async with self._lock:
            cached_entry = self._cache.get(key)

            if cached_entry is not None:
                self.cache_hits += 1
                return cached_entry

            self.cache_misses += 1
            return None

    async def add_async(
        self,
        *,
        resource_type: str,
        resource_id: str,
        bundle_entry: Optional[FhirBundleEntry],
        status: int,
    ) -> bool:
        """
        This method adds the given data to the cache.

        :param resource_type: The resource type to add the cached data for.
        :param resource_id: The resource id to add the cached data for.
        :param status: The status code of the request.
        :param bundle_entry: The cached data to add.
        :return: True if the entry was added, False if it already exists.
        """
        key: str = f"{resource_type}/{resource_id}"

        async with self._lock:
            # Check if the key already exists
            if key in self._cache:
                return False

            # Create the cache entry
            cache_entry = RequestCacheEntry(
                id_=resource_id,
                resource_type=resource_type,
                status=status,
                bundle_entry=bundle_entry,
            )

            # Add to the weak value dictionary
            self._cache[key] = cache_entry

            return True

    async def clear_async(self) -> None:
        """
        This method clears the cache.
        """
        async with self._lock:
            self._cache.clear()
