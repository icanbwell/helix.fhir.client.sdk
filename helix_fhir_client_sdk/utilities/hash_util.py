import hashlib


class ResourceHash:
    def __init__(self, hash_algorithm: str = "sha256") -> None:
        # Set the hash algorithm
        self.hash_algorithm = hash_algorithm.lower()

    def hash_value(self, value: str) -> str:
        # Hash a value using the specified algorithm
        if self.hash_algorithm not in hashlib.algorithms_available:
            raise ValueError(f"Hash algorithm {self.hash_algorithm} is not supported.")
        hasher = hashlib.new(self.hash_algorithm)
        hasher.update(value.encode("utf-8"))
        return hasher.hexdigest()
