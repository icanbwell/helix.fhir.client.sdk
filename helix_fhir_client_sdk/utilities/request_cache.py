from types import TracebackType
from typing import Dict, Optional, Type

from helix_fhir_client_sdk.fhir_bundle import BundleEntry


class RequestCache:
    """
    This is a class that caches requests to the FHIR server. It is used to avoid multiple requests to the FHIR server
    when we are doing a large number of requests for the same resource. It is used in the following way:

    """

    def __enter__(self) -> "RequestCache":
        """
        This method is called when the RequestCache is entered into a context manager. It returns the RequestCache
        instance.
        """
        self._cache: Dict[str, BundleEntry] = {}
        return self

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_value: Optional[BaseException],
        traceback: Optional[TracebackType],
    ) -> Optional[bool]:
        """
        This method is called when the RequestCache is exited from a context manager. It clears the cache.
        """
        self._cache.clear()
        if exc_type is not None:
            print(f"An exception of type {exc_type} occurred with message {exc_value}")
            return False  # Propagate any exception that occurred
        else:
            return True

    def __init__(self) -> None:
        self.cache_hits: int = 0
        self.cache_misses: int = 0

    def get(self, *, resource_type: str, resource_id: str) -> Optional[BundleEntry]:
        """
        This method returns the cached data for a given URL, or None if the URL is not in the cache.


        :param resource_type: The resource type to get the cached data for.
        :param resource_id: The resource id to get the cached data for.
        :return: The cached data for the given resource type and resource id, or None if the data is not in the cache.
        """
        key: str = f"{resource_type}/{resource_id}"
        if key in self._cache:
            self.cache_hits += 1
        else:
            self.cache_misses += 1
        return self._cache.get(key)

    def add(
        self, *, resource_type: str, resource_id: str, bundle_entry: BundleEntry
    ) -> None:
        """
        This method adds the given data to the cache.


        :param resource_type: The resource type to add the cached data for.
        :param resource_id: The resource id to add the cached data for.
        :param bundle_entry: The cached data to add.
        """
        key: str = f"{resource_type}/{resource_id}"
        self._cache[key] = bundle_entry

    def clear(self) -> None:
        """
        This method clears the cache.
        """
        self._cache.clear()
