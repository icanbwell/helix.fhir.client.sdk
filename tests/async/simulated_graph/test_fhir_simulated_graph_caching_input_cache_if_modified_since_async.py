import json
from datetime import UTC, datetime
from logging import Logger
from pathlib import Path
from typing import Any

from mockserver_client.mockserver_client import (
    MockServerFriendlyClient,
    mock_request,
    mock_response,
    times,
)

from helix_fhir_client_sdk.fhir_client import FhirClient
from helix_fhir_client_sdk.responses.fhir_get_response import FhirGetResponse
from helix_fhir_client_sdk.utilities.cache.request_cache import RequestCache
from helix_fhir_client_sdk.utilities.hash_util import ResourceHash
from tests.logger_for_test import LoggerForTest


async def test_fhir_simulated_graph_caching_input_cache_if_modified_since_async() -> None:
    logger: Logger = LoggerForTest()
    data_dir: Path = Path(__file__).parent.joinpath("./")
    graph_json: dict[str, Any]
    with open(data_dir.joinpath("graphs").joinpath("provider_if_modified_since.json")) as file:
        contents = file.read()
        graph_json = json.loads(contents)

    test_name = test_fhir_simulated_graph_caching_input_cache_if_modified_since_async.__name__

    mock_server_url = "http://mock-server:1080"
    mock_client: MockServerFriendlyClient = MockServerFriendlyClient(base_url=mock_server_url)

    relative_url: str = test_name
    absolute_url: str = mock_server_url + "/" + test_name

    mock_client.clear(f"/{test_name}/*.*")
    mock_client.reset()

    response_text: dict[str, Any] = {
        "resourceType": "Patient",
        "id": "1",
        "generalPractitioner": [{"reference": "Practitioner/5"}],
        "managingOrganization": {"reference": "Organization/6"},
    }

    mock_client.expect(
        request=mock_request(path=f"/{relative_url}/Patient/1", method="GET"),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {"resourceType": "Practitioner", "id": "5"}
    mock_client.expect(
        request=mock_request(path=f"/{relative_url}/Practitioner/5", method="GET"),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {"resourceType": "Organization", "id": "6"}
    mock_client.expect(
        request=mock_request(path=f"/{relative_url}/Organization/6", method="GET"),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {"entry": [{"resource": {"resourceType": "Coverage", "id": "7"}}]}
    mock_client.expect(
        request=mock_request(
            path=f"/{relative_url}/Coverage",
            querystring={"patient": "1"},
            method="GET",
        ),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {"entry": [{"resource": {"resourceType": "Observation", "id": "8"}}]}
    mock_client.expect(
        request=mock_request(
            path=f"/{relative_url}/Observation",
            querystring={
                "patient": "1",
                "category": "vital-signs,social-history,laboratory",
                "date": "ge2023-10-01",
            },
            method="GET",
        ),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {
        "resourceType": "Bundle",
        "total": 1,
        "type": "collection",
        "entry": [
            {
                "resource": {
                    "resourceType": "DocumentReference",
                    "id": "11",
                    "content": [
                        {"attachment": {"url": "Binary/12"}},
                        {"attachment": {"url": "Binary/13"}},
                        {"attachment": {"url": "Binary/14"}},
                        {"attachment": {"url": "Binary/15"}},
                    ],
                }
            }
        ],
    }
    mock_client.expect(
        request=mock_request(
            path=f"/{relative_url}/DocumentReference",
            querystring={"patient": "1", "date": "ge2023-10-01"},
            method="GET",
        ),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    # Mock request to return 403 so that Binaries could be fetched one by one
    mock_client.expect(
        request=mock_request(
            path=f"/{relative_url}/Binary",
            querystring={
                "_id": "12,13",
            },
            method="GET",
        ),
        response=mock_response(code=403),
        timing=times(1),
    )

    response_text = {"entry": [{"resource": {"resourceType": "Binary", "id": "12"}}]}
    mock_client.expect(
        request=mock_request(path=f"/{relative_url}/Binary/12", method="GET"),
        response=mock_response(body=response_text),
        timing=times(1),
    )
    response_text = {"entry": [{"resource": {"resourceType": "Binary", "id": "13"}}]}
    mock_client.expect(
        request=mock_request(path=f"/{relative_url}/Binary/13", method="GET"),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {"entry": [{"resource": {"resourceType": "Binary", "id": "14"}}]}
    mock_client.expect(
        request=mock_request(path=f"/{relative_url}/Binary/14", method="GET"),
        response=mock_response(body=response_text),
        timing=times(1),
    )
    response_text = {"entry": [{"resource": {"resourceType": "Binary", "id": "15"}}]}
    mock_client.expect(
        request=mock_request(path=f"/{relative_url}/Binary/15", method="GET"),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    encounter_json = {
        "resourceType": "Encounter",
        "id": "10",
        "participant": [{"individual": {"reference": "Practitioner/12345"}}],
    }
    response_text = {
        "entry": [
            {
                "resource": {
                    "resourceType": "Encounter",
                    "id": "8",
                    "participant": [{"individual": {"reference": "Practitioner/12345"}}],
                }
            },
            {"resource": encounter_json},
        ]
    }
    mock_client.expect(
        request=mock_request(
            path=f"/{relative_url}/Encounter",
            querystring={
                "patient": "1",
                "recorded-date": "ge2023-10-01",
            },
            method="GET",
        ),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    response_text = {"resourceType": "Practitioner", "id": "12345"}
    mock_client.expect(
        request=mock_request(
            path=f"/{relative_url}/Practitioner/12345",
            method="GET",
        ),
        response=mock_response(body=response_text),
        timing=times(1),
    )

    logger = LoggerForTest()
    fhir_client = FhirClient()
    fhir_client = fhir_client.expand_fhir_bundle(False)
    fhir_client.logger(logger=logger)

    fhir_client.extra_context_to_return({"service_slug": "medstar"})

    auth_access_token = "my_access_token"
    if auth_access_token:
        fhir_client = fhir_client.set_access_token(auth_access_token)

    request_cache = RequestCache(clear_cache_at_the_end=False)

    await request_cache.add_async(
        resource_type="Encounter",
        resource_id="10",
        from_input_cache=True,
        bundle_entry=None,
        status=200,
        last_modified=datetime.now(UTC),
        etag=None,
        raw_hash=ResourceHash().hash_value(json.dumps(encounter_json, sort_keys=True)),
    )

    fhir_client = fhir_client.url(absolute_url).resource("Patient")
    response: FhirGetResponse | None = await FhirGetResponse.from_async_generator(
        fhir_client.simulate_graph_streaming_async(
            id_="1",
            graph_json=graph_json,
            contained=False,
            separate_bundle_resources=False,
            request_size=2,
            input_cache=request_cache,
            ifModifiedSince=datetime(2023, 10, 1, 0, 0, 0, tzinfo=UTC),
        )
    )
    assert response is not None
    text = response.get_response_text()
    logger.info(text)

    logger.info(f"---------- Request Cache ({len(request_cache)}) ----------")
    async for entry in request_cache.get_entries_async():
        logger.info(entry)
    logger.info("---------- End Request Cache ----------")

    expected_file_path = data_dir.joinpath("expected")
    with open(expected_file_path.joinpath(test_name + ".json")) as f:
        expected_json: dict[str, Any] = json.load(f)

    bundle = json.loads(text)
    bundle["entry"] = [e for e in bundle["entry"] if e["resource"]["resourceType"] != "OperationOutcome"]

    # diff = DeepDiff(bundle, expected_json, ignore_order=True)
    # assert not diff, f"Diff: {diff}"

    # sort the entries by request url
    bundle["entry"] = sorted(bundle["entry"], key=lambda x: int(x["resource"]["id"]))
    bundle["total"] = len(bundle["entry"])
    logger.info(f"-------- Actual Bundle ({len(bundle['entry'])}) --------")
    for entry in bundle["entry"]:
        logger.info(entry)
    logger.info("-------- End Actual Bundle --------")
    logger.info(f"-------- Expected Bundle ({len(expected_json['entry'])}) --------")
    for entry in expected_json["entry"]:
        logger.info(entry)
    logger.info("-------- End Expected Bundle --------")
    assert bundle == expected_json
