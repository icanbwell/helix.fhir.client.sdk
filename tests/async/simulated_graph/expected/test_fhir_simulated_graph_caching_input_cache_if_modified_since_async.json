{
  "resourceType": "Bundle",
  "total": 10,
  "type": "collection",
  "entry": [
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/Patient/1"
      },
      "resource": {
        "generalPractitioner": [
          {
            "reference": "Practitioner/5"
          }
        ],
        "id": "1",
        "managingOrganization": {
          "reference": "Organization/6"
        },
        "resourceType": "Patient"
      },
      "response": {
        "status": "200"
      }
    },
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/Practitioner/5"
      },
      "resource": {
        "id": "5",
        "resourceType": "Practitioner"
      },
      "response": {
        "status": "200"
      }
    },
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/Organization/6"
      },
      "resource": {
        "id": "6",
        "resourceType": "Organization"
      },
      "response": {
        "status": "200"
      }
    },
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/Encounter?patient=1&recorded-date=ge2023-10-01T00:00:00+00:00"
      },
      "resource": {
        "id": "8",
        "participant": [
          {
            "individual": {
              "reference": "Practitioner/12345"
            }
          }
        ],
        "resourceType": "Encounter"
      },
      "response": {
        "status": "200"
      }
    },
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/DocumentReference?patient=1&date=ge2023-10-01T00:00:00+00:00"
      },
      "resource": {
        "id": "11",
        "resourceType": "DocumentReference",
        "content": [
          {
            "attachment": {
              "url": "Binary/12"
            }
          },
          {
            "attachment": {
              "url": "Binary/13"
            }
          },
          {
            "attachment": {
              "url": "Binary/14"
            }
          },
          {
            "attachment": {
              "url": "Binary/15"
            }
          }
        ]
      },
      "response": {
        "status": "200"
      }
    },
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/Binary/12"
      },
      "resource": {
        "id": "12",
        "resourceType": "Binary"
      },
      "response": {
        "status": "200"
      }
    },
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/Binary/13"
      },
      "resource": {
        "id": "13",
        "resourceType": "Binary"
      },
      "response": {
        "status": "200"
      }
    },
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/Binary/14"
      },
      "resource": {
        "id": "14",
        "resourceType": "Binary"
      },
      "response": {
        "status": "200"
      }
    },
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/Binary/15"
      },
      "resource": {
        "id": "15",
        "resourceType": "Binary"
      },
      "response": {
        "status": "200"
      }
    },
    {
      "request": {
        "method": "GET",
        "url": "http://mock-server:1080/test_fhir_simulated_graph_caching_input_cache_if_modified_since_async/Practitioner/12345"
      },
      "resource": {
        "id": "12345",
        "resourceType": "Practitioner"
      },
      "response": {
        "status": "200"
      }
    }
  ]
}
